bpfd/src/bpf.rs

```diff
@@ -1143,134 +908,154 @@ impl BpfManager {
     ) -> Result<(), BpfdError> {
-        let (_, map_index) = get_map_index(id, map_owner_uuid);
-
-        if let Some(map) = self.maps.get_mut(&map_index.clone()) {
-            if let Some(index) = map.used_by.iter().position(|value| *value == id) {
-                map.used_by.swap_remove(index);
-            }
+        // returns the id to be used for maps, it could be the program id OR another program's id.
+        let id = map_owner_id.unwrap_or(prog_id);
+        // let path = PathBuf::from(format!("{RTDIR_FS_MAPS}/{}", id));
+        if let Some(mut map) = self.maps.remove(&id) {
+            map.used_by.retain(|value| *value != prog_id);
 
             if map.used_by.is_empty() {
-                let (_, path) = calc_map_pin_path(id, map_owner_uuid);
-                self.maps.remove(&map_index.clone());
-                fs::remove_dir_all(path)
+                fs::remove_dir_all(map.clone().map_pin_path)
                     .await
                     .map_err(|e| BpfdError::Error(format!("can't delete map dir: {e}")))?;
+            } else {
+                self.maps.insert(id, map);
             }
         } else {
-            return Err(BpfdError::Error("map_pin_path does not exists".to_string()));
+            return Err(BpfdError::Error(
+                "map_owner_id does not exist cannot delete".to_string(),
+            ));
         }
 
         Ok(())
     }
 
-    fn rebuild_map_entry(&mut self, id: Uuid, map_owner_uuid: Option<Uuid>) {
-        let (_, map_index) = get_map_index(id, map_owner_uuid);
+    fn rebuild_map_entry(&mut self, prog_id: u32, map_owner_id: Option<u32>) {
+        let id = map_owner_id.unwrap_or(prog_id);
+        let path = PathBuf::from(format!("{RTDIR_FS_MAPS}/{}", id));
 
-        if let Some(map) = self.maps.get_mut(&map_index) {
-            map.used_by.push(id);
+        if let Some(map) = self.maps.get_mut(&id) {
+            map.used_by.push(prog_id);
         } else {
-            let (_, map_pin_path) = calc_map_pin_path(id, map_owner_uuid);
             let map = BpfMap {
-                map_pin_path: map_pin_path.clone(),
-                used_by: vec![id],
+                map_pin_path: path.clone(),
+                used_by: vec![prog_id],
             };
-            self.maps.insert(map_index, map);
+            self.maps.insert(id, map);
         }
     }
 }
 
-// map_index is a UUID. It is either the programs UUID, or the UUID
-// of another program that map_owner_uuid references.
-// This function also returns a bool, which indicates if the input UUID
-// is the owner of the map (map_owner_uuid is not set) or not (map_owner_uuid
-// is set so the eBPF program is referencing another eBPF programs maps).
-fn get_map_index(id: Uuid, map_owner_uuid: Option<Uuid>) -> (bool, Uuid) {
-    if let Some(uuid) = map_owner_uuid {
-        (false, uuid)
-    } else {
-        (true, id)
+fn sort_dispatcher_programs(mut programs: Vec<Program>) -> Vec<Program> {
```

Comment:
This is not the algorithm we use to sort programs :(
See [here](
https://github.com/xdp-project/xdp-tools/blob/master/lib/libxdp/protocol.org#loading-and-attaching-component-programs) where our previous sort relied on the `Metadata` struct which contained the name + whether a program was already loaded or not. It's PartialEq impl, used by sort, did the hard work for us. Size, tag and load_time should also have been added.

Reply:
That links to an empty playground, you needed the permalink 😉 
I'm not advocating that you do the work of fixing the sort to be libxdp compatible - that's on me.
I'd only ask that you don't change the sorting algorithm in this PR and maintain the old behaviour.